Upgrade Hapi in legacy platform to v17

joshdover opened the pull request:
<!--
Thank you for your interest in and contributing to Kibana! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md)?
- If submitting code, have you included unit tests that cover the changes?
- If submitting code, have you tested and built your code locally prior to submission with `yarn test && yarn build`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
-->

Fixes #13802 
Fixes https://github.com/elastic/kibana/issues/21140

### Summary of changes
- Removed even-better fork of good
- Updated all route handlers to use async/await rather than the `reply` callback
- Updated changes to `server.auth` calls
- Upgrade hapi, joi, wreck, h2o2, hapi-auth-cookie, inert, vision, boom
- Added new validation `failAction` handler that returns errors similar to Hapi v14
- Changes to how we use exports from joi
- Reimplemented monitoring built on good. This time around, I'm going to push to get it accepted into the mainline package so we don't have to maintain a fork. PR here: https://github.com/hapijs/good/pull/593
- Made subtle changes to how security plugin interacts with Hapi

### Risks

As with any major upgrade, this PR has risks for breakage across the entire application. While I have made sure to pay attention to all these areas, it's possible I've missed things. Areas we may want to focus on in QA:
- **Validation errors.** Any features that depend on the backend telling us that the frontend sent valid data should be checked. In particular how the error message is displayed may be broken in some areas. I believe I was able to get this to be exactly the same before, but it's possible some subtle thing was missed.
- **Authentication.** This upgrade changes how we tell Hapi who the authenticated user is. We should validate that nothing about this behavior is wrong.
- **Logging.** I have upgraded our core logging components in the legacy platform. I don't anticipate any major problems here.
- **Kibana Monitoring.** For similar reasons, our monitoring of Kibana itself may have changed slightly. I have done my best to verify that we're sending the same data to ES for kibana's internal monitoring, but it's possible it's not quite right or misbehaves in bad conditions.

### Potential Areas for Improvement
- In order to have the same validation errors return from our backend as we did in Hapi v14, we have to install a default validation `failAction` handler. This means that any tests that don't set up their test server with this handler will not have the same validation responses as the real environment does. It would be a nice improvement if we had a single test server fixture that we used everywhere.

# Release notes

release-note: The Hapi framework has been upgraded from v14.2.0 to v17.5.3. There are a number of breaking changes affecting plugins that register custom backend routes. The `reply` callback interface for supplying responses has been replaced with an async/await interface and a response toolkit. More details in the [Hapi upgrade guide](https://github.com/hapijs/hapi/issues/3658)

# Dev Docs

## Upgraded Hapi framework for plugin backend endpoints
The Hapi framework has been upgraded from v14.2.0 to v17.5.3. There are a number of breaking changes affecting plugins that register custom bacekend routes. This does not affect plugins that do not have backend routes.

The `reply` callback interface for supplying responses has been replaced with an async/await interface and a "response toolkit". Example:

### Before 6.5
```js
server.route({
  path: '/myroute',
  handler(req, reply) {
    getSomeData().then((data) => reply(data));
  }
});

server.route({
  path: '/myredirect',
  handler(req, reply) {
    reply.redirect('/otherroute');
  }
});
```

### After 6.5
```js
server.route({
  path: '/myroute',
  async handler(req, h) {
    return await getSomeData();
  }
});

server.route({
  path: '/myredirect',
  handler(req, h) {
    return h.redirect('/otherroute');
  }
});
```

More details can be found in the [Hapi upgrade guide](https://github.com/hapijs/hapi/issues/3658).

tylersmalley commented on src/cli/serve/serve.js:
```diff
@@ -250,9 +250,9 @@ export default function (program) {
-        kbnServer.server.log(['info', 'config'], 'Reloading logging configuration due to SIGHUP.');
-        await kbnServer.applyLoggingConfiguration(config);
-        kbnServer.server.log(['info', 'config'], 'Reloaded logging configuration due to SIGHUP.');
+        // kbnServer.server.log(['info', 'config'], 'Reloading logging configuration due to SIGHUP.');
```
Oh yeah, this is all coming back to me. We need to get off our fork at https://github.com/bevacqua/even-better/commit/4342ed5549e70ac46b34f93ade2b5d067cef30d5

joshdover replied:
Yep, I got some good details from Court on this. Going to go with getting everything working and then re-implement the things that depend on good with either a new version or something else.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8923/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8983/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/9173/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/9210/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/14/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/22/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/11/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/287/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/355/)

joshdover commented on src/core/server/legacy_compat/logging/legacy_logging_server.ts:
```diff
@@ -17,20 +17,20 @@
  */
 
-import { EventEmitter } from 'events';
+import { ServerExtType } from 'hapi-latest';
```
@azasypkin May want to take a look at the changes I'm making here.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/384/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/445/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/486/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/491/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/499/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/504/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/555/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/564/)

joshdover commented on src/core/server/legacy_compat/legacy_platform_proxifier.ts:
```diff
@@ -106,9 +106,8 @@ export class LegacyPlatformProxifier extends EventEmitter {
-  public getConnections(callback: (error: Error | null, count?: number) => void) {
-    // This method is used by `even-better` (before we start platform).
-    // It seems that the latest version of parent `good` doesn't use this anymore.
+  public getConnections(callback: (error: Error | null, count?: number) => number) {
```
This is just passing through to [Node's net.Server.getConnections method](https://nodejs.org/api/net.html#net_server_getconnections_callback). It gives us the number of active connections on the server. We're currently showing this value on the monitoring dashboard in Kibana.

azasypkin replied:
Yeah, but that's `count?` in this signature `(error: Error | null, count?: number) => number`, what about return value `=> number`?

joshdover replied:
Ah I see what you mean, I changed this quite a few times. Need to put it back to what it was, thanks for catching!

joshdover commented on src/core/server/legacy_compat/logging/legacy_logging_server.ts:
```diff
@@ -41,11 +41,20 @@ interface PluginRegisterParams {
   constructor(legacyLoggingConfig: Readonly<Record<string, any>>) {
-    super();
+    // Tell our podium instance about the events we'll be using.
+    this.events.registerEvent('log');
```
Hapi replaced it's event bus with Podium (previously was an EventEmitter). This library requires that events that are emitted are first published. More deets here: https://github.com/hapijs/hapi/issues/3323

azasypkin replied:
Just played with this PR and dug into `good` source code and I think we can simplify this file a bit since `LegacyLoggingServer` shouldn't log any `ops` info so it can [just ignore `onPreStart` completely](https://github.com/hapijs/good/blob/master/lib/index.js#L17-L24). What about something like this?
```ts
export class LegacyLoggingServer {
  // Tell our podium instance about the events we'll be using. We don't actually
  // forward `request` and `response` events, but we have to register them with Podium
  // in order for the plugin to be able to setup.
  public events: Podium = new Podium(['log', 'request', 'response']);

  private onStopCallback?: () => void;

  constructor(legacyLoggingConfig: Readonly<Record<string, any>>) {
    setupLogging(
      this,
      Config.withDefaultSchema(transformDeprecations({ logging: legacyLoggingConfig }))
    );
  }
  .....
  public log({ level, context, message, error, timestamp, meta = {} }: LogRecord) {
    this.events.emit('log', { ... });
  }

  public stop() {
    if (this.onStopCallback !== undefined) {
      this.onStopCallback();
    }
  }

  public ext(eventName: ServerExtType, callback: () => void) {
    // We care only about stop callback here.
    if (eventName === 'onPostStop') {
      this.onStopCallback = callback;
    }
  }

  public expose() {....}
}

```

joshdover replied:
Good point, I'll simplify this. Originally I was thinking about making this resilient to any changes we made to good, but I don't expect any and think it's pretty likely we'll drop it in the near future. FYI, we are using our fork of good at elastic/good.

joshdover commented on src/core/server/legacy_compat/logging/legacy_logging_server.ts:
```diff
@@ -61,23 +70,38 @@ export class LegacyLoggingServer extends EventEmitter {
+    this.onPostStopCallbacks.forEach(callback => callback());
+  }
+
+  public ext(eventName: ServerExtType, callback: () => void) {
```
Logging will never begin if we don't call on the onPreStart. onPreStop closes any logging streams which in this case could be closing a file.

azasypkin replied:
So if we `this.events.emit('log',....)` without calling `onPreStart` nothing will happen, right?

joshdover replied:
Right if we don't call onPreStart then the logging plugin won't be listening for those events.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/637/)

joshdover commented on src/server/http/index.js:
```diff
@@ -39,28 +37,40 @@ export default async function (kbnServer, server, config) {
     state: {
       strictHeader: false,
+
+      // Remain compat with Hapi v14
```
They're to emulate the previous behavior that was changed in [Hapi v15](https://github.com/hapijs/hapi/issues/3323). We could investigate more here to see if we're already overriding these defaults.

azasypkin replied:
Okay, thanks. But don't forget to mirror any additional settings in `src/core/server/http/http_tools.ts` before you merge this PR.

joshdover replied:
I think it'd actually be better if we can drop these, as they're definitely less secure. As far as I can tell, using the more secure defaults work fine on my local instance. Wondering if there's any weird edges/use cases that others might know of that I don't (specifically with cross-domain cookies or client-side cookie reading/modification).

@elastic/kibana-security Can you think of any reason NOT to use the `SameSite`, `Secure`, and `HttpOnly` flags on cookies?

kobelb replied:
We're currently setting the `HttpOnly` flag on the session cookie, and when Kibana itself is using SSL (and they aren't doing SSL termination using a reverse proxy) we're setting the `Secure` flag as well.

I don't think we'll want to set `SameSite` to `Lax` or `Strict` by default, given that we have users embedding Kibana in cross origin iframes, if we do so then the user will be required to login every time that they change pages on the embedding site. I'd love to see the ability to easily opt into `X-Frame-Options: sameorigin` and `SameSite` cookies though, and there's the possibility that we could change this to be the default behavior in a major with the ability to opt-out of it, but I don't think that should be in scope for this PR.

joshdover commented on src/server/http/secure_options.test.js:
```diff
@@ -17,30 +17,31 @@
 describe('secure_options', function () {
   it('allows null', function () {
-    expect(secureOptions(null)).toBe(null);
+    expect(secureOptions(null)).to.be(null);
```
Ah yes meant to put this back. Was using the wrong test runner.

joshdover commented on src/server/config/__tests__/deprecation_warnings.js:
```diff
@@ -20,6 +20,7 @@
 import { spawn } from 'child_process';
 
 import expect from 'expect.js';
+import { race, delay } from 'bluebird';
```
Good to know! I'll swap it out!

joshdover commented on src/cli/cluster/configure_base_path_proxy.test.js:
```diff
@@ -25,7 +25,6 @@ jest.mock('../../server/logging', () => ({
   setupLogging: jest.fn(),
 }));
```
Great!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/659/)

epixa commented:
@azasypkin Do we need to do anything in the new platform to accommodate logging changes?

@elastic/kibana-security Heads up on this since it impacts how authentication integrates with hapi, you might want to take a look.

@elastic/kibana-qa Heads up in case you start seeing Weird Stuff™ happening after this gets merged.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/674/)

kobelb commented on x-pack/plugins/security/server/routes/api/public/roles/put.test.js:
```diff
@@ -33,7 +41,7 @@ const putRoleTest = (
     const mockPreCheckLicense = jest
       .fn()
-      .mockImplementation(preCheckLicenseImpl);
+      .mockImplementation(preCheckLicenseImpl || defaultPreCheckLicenseImpl);
```
Would you mind explaining why this is necessary and why the check for the `mockPreCheckLicense` not being called in the else below was removed?

joshdover replied:
This is a good catch and something I meant to rollback after I had finished fixing this code up. I'm going to do another read pass through my test changes to make sure I didn't miss anything else.

kobelb commented on x-pack/plugins/security/server/routes/api/v1/__tests__/authenticate.js:
```diff
@@ -142,14 +123,12 @@ describe('Authentication routes', () => {
-        sinon.assert.notCalled(replyStub);
-        sinon.assert.calledOnce(replyStub.continue);
-        sinon.assert.calledWithExactly(replyStub.continue, { credentials: user });
+        sinon.assert.calledOnce(hStub.response);
```
Is there a way to not have to stub out the `h` response toolkit, and instead do the asserts only based on the response value of the handler? If I understand correctly, the return values from the handlers should contain all of the information that we need if we could pass an actual implementation of `h` to the handlers. Not sure how Hapi 17 exposes this for tests like this.

I don't have any real issue with the way that you've done this here, just asking for future tests.

joshdover replied:
In a test like this where we're using a server stub rather than an actual Hapi.Server instance, I couldn't find an _officially supported_ way to create a fake toolkit for the purpose of testing. Now we could use a real server instance and use `await server.inject` and then inspect it's return value.

I think testing like that, while it may be slower, does allow our tests to only verify the output rather than the implementation details. It would make refactors/package upgrades like this require many fewer test changes.

azasypkin replied:
>  Now we could use a real server instance and use await server.inject and then inspect it's return value.

> I think testing like that, while it may be slower, does allow our tests to only verify the output rather than the implementation details. It would make refactors/package upgrades like this require many fewer test changes.

Just wanted to note that we should not rely on `server.inject` for two main reasons:

1. In master all requests go through new platform/core, so bypassing core in tests with `kbnServer.server.inject` won't test real setup
2. We should _try_ to abstract away from Hapi specific APIs as much as possible since we would like to be able to switch to another HTTP framework in the future and tests should help us here, not slow us down

For that purpose I'm changing [`test_utils/kbn_server`](https://github.com/elastic/kibana/pull/22190/files#diff-89a18ea56e901a28a29e99fe9a697823) so that we can use `supertest` instead of `HapiServer.inject` [like that](https://github.com/elastic/kibana/pull/22190/files#diff-28b96eccb002d27f70dfeb7907bca289R45) that will work with any wrapper around native `http.Server` (ExpressJS, Hapi, Koa etc.). It still uses some hackery inside to access core's private properties, but it's hidden from consumers and we'll improve that at some point.

joshdover replied:
Ah I like this a lot. Sounds like a great step forward. Hadn't thought about it, but I agree getting away from framework-specifics is better. Sounds like that in the time being, leaving this particular test as-is until we can use `supertest` is probably best.

chrisronline commented on x-pack/plugins/monitoring/server/kibana_monitoring/collectors/get_ops_stats_collector.js:
```diff
@@ -4,48 +4,74 @@
-      monitor.removeListener('ops', onOps);
-    }
-  };
+  const interval = kbnServer.config.get('ops.interval');
```
Just want to double check with @tsullivan here.

We are currently already using this same config value to configure how often we fetch ops data from hapi right? 

[Here](https://github.com/elastic/kibana/blob/master/src/server/logging/configuration.js#L55) from [here]()https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/kibana_monitoring/collectors/get_ops_stats_collector.js#L19) from [here](https://github.com/elastic/kibana/blob/master/src/server/kbn_server.js#L182)

tsullivan replied:
This config value is supposed to define the granularity of the ops event emitter, which means how often the event fires, and how large the window is for including stats. It looks like this PR isn't changing that

tsullivan commented on package.json:
```diff
@@ -100,25 +100,24 @@
     "font-awesome": "4.4.0",
     "glob": "^7.1.2",
     "glob-all": "^3.1.0",
+    "good": "github:joshdover/good#b9adb7f4e90bca43ef6c01bfcbd9359481a93bbc",
```
Will this be published before this gets merged?

joshdover replied:
I have an email thread with the maintainer going. He is needing some convincing to accept our changes, so as of now, I'm not 100% sure. Happy to move this / publish to NPM as need to get this live.

tsullivan commented on src/server/status/lib/metrics.js:
```diff
@@ -98,8 +104,11 @@ export class Metrics {
+        ...requestDefaults,
+        ...keysToSnakeCaseShallow(get(hapiEvent, ['requests', port]))
+      },
+      concurrent_connections: hapiEvent.concurrent_connections
```
Something seems to be up with capturing requests stats from ops event data. I'm not seeing the requests stats ever count anything
![image](https://user-images.githubusercontent.com/908371/44550614-9f911680-a6d9-11e8-8cc3-2a309040b34b.png)

![image](https://user-images.githubusercontent.com/908371/44550698-d9fab380-a6d9-11e8-870d-c580628e7562.png)

joshdover replied:
Thanks for taking a looking and helping me catch this. I'll ping you when I have a fix.

joshdover replied:
This was a fun one. Turns out the library the official hapi library for getting this data (hapijs/oppsy) emits events with objects that it then later resets. This was just a race condition with also getting the number of connections on the server. I've changed this to do a deep copy of the event data, but I'm also going to submit a patch upstream.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/796/)

tylersmalley commented on docs/development/core/development-basepath.asciidoc:
```diff
@@ -55,8 +55,8 @@ Append `config.get('server.basePath')` to any absolute URL path.
-  handler(req, reply) {
-    reply.redirect(`${basePath}/otherLocation`);
+  handler(req, h) {
+    return h.redirect(`${basePath}/otherLocation`);
```
For examples, I would prefer to keep the long-form variable names.

```
  handler(req, reply) {
    return reply.redirect(`${basePath}/otherLocation`);
  }
```

joshdover replied:
Hapi convention is to call this `h` now. They've removed the `reply` name completely. I agree it's not a great choice, but might be easier for us if we're consistent with hapi documentation?

tylersmalley replied:
Ok, I think this is fine for the old platform. In the new platform we are not interfacing with the entire router as to not expose any router internals, which is still Hapi in the new platform.

joshdover commented:
@tsullivan Should be good to check out again. Found and fixed the race condition with that lib. I fixed it from our end but also submitted a PR: hapijs/oppsy#6. We could drop that deep copy in our code once/if it's accepted.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/832/)

epixa commented:
@joshdover I know I told you that we should target a branch on the good fork, but I actually think we should target a tag instead. We should tag it with something like `<hapi version>-kibana1`. This will make it clear exactly what we're targeting from the package.json, and it'll allow us to bump the right side if we make additional changes or the left side if we merge upstream changes in.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/881/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/883/)

tylersmalley commented on src/server/saved_objects/routes/delete.test.js:
```diff
@@ -22,7 +22,7 @@ import { createDeleteRoute } from './delete';
 
 describe('DELETE /api/saved_objects/{type}/{id}', () => {
-  const savedObjectsClient = { delete: sinon.stub() };
+  const savedObjectsClient = { delete: sinon.stub().returns('') };
```
Why an empty string? SavedObjectClient.delete returns a promise.

tylersmalley commented on src/server/status/index.js:
```diff
@@ -21,6 +21,8 @@ import ServerStatus from './server_status';
 import { Metrics } from './lib/metrics';
 import { registerStatusPage, registerStatusApi, registerStatsApi } from './routes';
 import { getOpsStatsCollector } from './collectors';
+import Oppsy from 'oppsy';
```
This needs to be added as a dependency if we are directly using it.

joshdover replied:
Ah good call.

tylersmalley commented on package.json:
```diff
@@ -100,25 +100,24 @@
     "handlebars": "4.0.5",
-    "hapi": "14.2.0",
-    "hapi-latest": "npm:hapi@17.5.0",
+    "hapi": "17.5.3",
```
Let's go ahead and use carets for the version matching here.

joshdover replied:
SGTM

tylersmalley commented on x-pack/plugins/apm/server/routes/errors.js:
```diff
@@ -43,9 +43,7 @@ export function initErrorsApi(server) {
-      })
-        .then(reply)
-        .catch(defaultErrorHandler(reply));
+      }).catch(defaultErrorHandler);
```
Did we intentionally drop the response here?

tylersmalley commented on x-pack/plugins/grokdebugger/server/lib/license_pre_routing_factory/__tests__/license_pre_routing_factory.js:
```diff
@@ -34,32 +34,28 @@ describe('license_pre_routing_factory', () => {
-          expect(response).to.eql(Boom.forbidden());
-          done();
-        });
+        const response = licensePreRouting(stubRequest);
```
This needs to be awaited.

joshdover replied:
The factory is now a regular function, not async. It only used callbacks before b/c of the reply function, not because what it looksup is asynchronous.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1112/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1121/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1129/)

spalger commented on src/server/http/max_payload_size.test.js:
```diff
@@ -65,6 +65,6 @@ test('fails with 400 if payload size is larger than default and route config all
   });
 
-  expect(resp.statusCode).toBe(400);
+  expect(resp.statusCode).toBe(413);
```
This seems like a potentially unrelated change that should be in a separate PR if it's really necessary.

joshdover replied:
Hapi changed how max body size errors were handled to return the correct error code. Talked to Tyler about this and we thought this would be a pretty safe change. I checked that client-side code does not break from this change.

spalger commented on src/server/http/register_hapi_plugins.js:
```diff
@@ -20,14 +20,11 @@
-  await fromNode(cb => {
-    server.register(plugins, cb);
-  });
+  return server.register(plugins);
```
Can we use `return await` here too?

joshdover replied:
Not sure what the benefit of that would be? The caller still has to `await` it as well.

spalger replied:
Yeah, personally I just always `return await` when that's what I want because I think it makes it a little clearer to readers down the road that `server.register()` is returning a promise that needs to be awaited, and it helps avoid the situation where someone else comes in and tries to wrap this in a `try {} catch {}` and runs into the problem we ran into in other parts of this PR because we weren't using `return await`.

joshdover replied:
Got it, yeah that makes sense. Thanks.

spalger commented on src/server/http/validation_error_handler.js:
```diff
@@ -0,0 +1,23 @@
+
+// Used to replicate Hapi v16 and below's validation responses
+export function defaultValidationErrorHandler(request, h, err) {
+  throw err;
```
Hmm, what's the default behavior in Hapi 17 that is different from 15? Is this possibly necessary because we are returning errors rather than throwing them?

joshdover replied:
No, Hapi stopped exposing the actual details of validation errors. This is the community recommended way of exposing these errors the same way as before.

spalger replied:
Ah, I missed that this is specifically for the `validate:` route config. 👍

joshdover replied:
I guess I didn't answer your original question: it doesn't matter whether your throw or return the Boom object.

joshdover commented on src/core_plugins/console/server/proxy_route.js:
```diff
@@ -73,21 +73,21 @@ export const createProxyRoute = ({
+          throw err;
         }
+
+        return null;
```
In Hapi v17+, you do have to return values from pre responses. An `undefined` value will actually make Hapi throw and respond with a 500.

Agree on reversing the if, I'll change that.

joshdover commented on src/core_plugins/console/index.js:
```diff
@@ -113,14 +113,13 @@ export default function (kibana) {
           if (!apis) {
-            reply(Boom.badRequest('"apis" is a required param.'));
-            return;
+            return Boom.badRequest('"apis" is a required param.');
```
I debated this a few times too. I opted for returning rather than throwing to make fewer test changes. I found this preferable during a big refactor because I think changing lots of tests introduces more risk to introducing an unintended behavior change. What do you think?

spalger replied:
Hmm, the tests behave differently when a route handler throws vs returns an error? How so?

joshdover replied:
Actually, I'm remembering this wrong I think. I know I tried doing this at first and I broke a lot of things. Let me try again and figure out what it broke.

joshdover replied:
It required some test changes to things that called handlers directly, but wasn't so bad. Have pushed a new commit.

joshdover commented on src/server/http/__snapshots__/max_payload_size.test.js.snap:
```diff
@@ -1,3 +1,3 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`fails with 400 if payload size is larger than default and route config allows 1`] = `"{\\"statusCode\\":400,\\"error\\":\\"Bad Request\\",\\"message\\":\\"Payload content length greater than maximum allowed: 200\\"}"`;
+exports[`fails with 400 if payload size is larger than default and route config allows 1`] = `"{\\"statusCode\\":413,\\"error\\":\\"Request Entity Too Large\\",\\"message\\":\\"Payload content length greater than maximum allowed: 200\\"}"`;
```
An unexpected change or something is wrong with the snapshot format?

spalger replied:
I mean the snapshot doesn't match the test description, so maybe it's just the test title that is out of date.

joshdover commented on src/server/config/schema.test.js:
```diff
@@ -51,20 +51,20 @@ describe('Config schema', function () {
         const { error } = validate({ server: { basePath: '/path/' } });
         expect(error).toHaveProperty('details');
-        expect(error.details[0]).toHaveProperty('path', 'server.basePath');
+        expect(error.details[0]).toHaveProperty('path', ['server', 'basePath']);
```
Yep, from Joi upgrade. I checked this earlier in this project, but let me do another scan through now that it's more complete.

joshdover commented on src/core_plugins/kibana/server/routes/api/scroll_search/index.js:
```diff
@@ -33,21 +33,26 @@ export function scrollSearchApi(server) {
-        .catch(error => reply(handleESError(error)));
+
+      try {
+        return callWithRequest(req, 'search', params);
```
That's a good point and something hadn't thought about. Will double check + fix all these.

joshdover commented on src/optimize/watch/watch_optimizer.js:
```diff
@@ -63,10 +63,9 @@ export default class WatchOptimizer extends BaseOptimizer {
-      this.onceBuildOutcome()
-        .then(() => reply.continue())
-        .catch(reply);
+    server.ext('onRequest', (request, h) => {
```
👍

joshdover commented on src/cli_plugin/install/downloaders/http.js:
```diff
@@ -41,32 +40,27 @@ function getProxyAgent(sourceUrl, logger) {
-  });
+    return { req, resp };
+  } catch (err) {
+    throw new Error('ENOTFOUND');
```
Oh right, good catch. This confirms my suspicion that changing everything from promises to async/await would be non-trivial/prone to mistakes.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1165/)

tylersmalley commented:
Talked with @joshdover - he is looking into shimming the `server.route` handler to support the legacy reply function. This would eliminate the breaking change, thus making it easier for us to reason about backporting to 6.x.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1234/)

spalger commented on src/core_plugins/kibana/server/routes/api/home/register_tutorials.js:
```diff
@@ -22,8 +22,8 @@ export function registerTutorials(server) {
-    handler: async function (req, reply) {
-      reply(server.getTutorials());
+    handler: async function () {
+      return server.getTutorials();
```
I'm not sure if `server.getTutorials()` returns a promise here, but if it was using `return await` I would know for sure, and if it's not then there's no need for this to be an `async` function.

joshdover commented:
We've decided to move forward with this breaking change. The shim ended up being quite complicated: https://gist.github.com/joshdover/959b7094f848a099cd2f1615ca8e7375

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1275/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1367/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1370/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1374/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1426/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1676/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2029/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2046/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2062/)

spalger commented:
@joshdover I have some more time to look at this, if you wouldn't mind resolving the conflicts with master

epixa commented on package.json:
```diff
@@ -100,25 +100,24 @@
     "font-awesome": "4.4.0",
     "glob": "^7.1.2",
     "glob-all": "^3.1.0",
+    "good": "github:elastic/good#v8.1.1-kibana1",
```
I know I suggested pulling in the dependency like this, but this won't work after all. Any dependency that is referenced via a github link/url (or any url) will cause yarn to reach out and resolve those packages every time. I guess the thought process is that yarn can't reliably determine the state of a package unless it can rely on the behaviors of the upstream service, and it hasn't built-in knowledge of github commit handling like it has npm.

Looks like we'll need to publish this to `@elastic/good` on npm instead. Ping me directly and we can chat details.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2713/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2740/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2759/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2779/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2826/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3181/)

tylersmalley commented:
Apologies for the delay on this - my code review was completed earlier, just needed to finish my functional testing. Everything looks good, though we are getting an unexpected log output.

>   log   [05:14:19.356] [error][client][connection] undefined

This is after I login and trigger a refresh

joshdover commented:
> > log   [05:14:19.356] [error][client][connection] undefined

Fixed this issue!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3334/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3354/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3370/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3390/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3536/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3792/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3797/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3897/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3902/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3950/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4055/)

jbudz commented on docs/migration/migrate_7_0.asciidoc:
```diff
@@ -72,3 +72,10 @@ considered unique based on its persistent UUID, which is written to the path.dat
+
+[float]
+=== Upgraded Hapi framework for plugin backend endpoints
+*Details:* The Hapi framework has been upgraded from v14.2.0 to v17.5.3. There are a number of breaking changes affecting plugins that register custom bacekend routes. The `reply` callback interface for supplying responses has been replaced with an async/await interface and a response toolkit. More details in the [Hapi upgrade] guide(https://github.com/hapijs/hapi/issues/3658)
```
s/bacekend/backend :)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4103/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4275/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4336/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4359/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4382/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4385/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4669/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4727/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4733/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4863/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4892/)

joshdover commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4923/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5108/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5127/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5133/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5163/)

joshdover commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5170/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5588/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5653/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5820/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6469/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6501/)

sqren commented on x-pack/plugins/apm/server/routes/transactions.ts:
```diff
@@ -18,18 +18,18 @@ import { getSpans } from '../lib/transactions/spans/get_spans';
   console.error(err.stack);
   // @ts-ignore
-  reply(Boom.wrap(err, err.statusCode || 400));
+  Boom.boomify(err, { statusCode: err.statusCode || 400 });
```
@joshdover Minor thing: this should be returned or thrown. We've created an issue for it here: https://github.com/elastic/kibana/issues/24844

We'll fix the APM instances (if there are more) but just a heads up if this is also the case for other plugins.
